### PR TITLE
Improve LRO polling

### LINKED ISSUE #914: Implement LRO polling mechanism based on ARM RPC
[Issue 1511889: Swagger Correctness: Microsoft.Compute/virtualMachines on 2021-05-01 - Boards (azure.com)](https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1511889)

Honor `x-ms-long-running-operation-options`, especially `final-state-via`
~polling url must not contain secret/token~ moved to #925

### PR DESCRIPTION
**Pull request by jianyexi**
fix https://github.com/Azure/oav/issues/914

### CODE REVIEW

File: lib/apiScenario/postmanCollectionRunnerClient.ts [hunk 1]
```diff
@@ -693,19 +704,49 @@ pm.test("Stopped TestProxy recording", function() {
...
   ) {
     if (this.opts.skipLroPoll) return;
-
+    const url = item.request.url;
+    const urlStr = `${baseUri}${url.getPathWithQuery()}`;
     postScripts.push(
       `
+ function getLroFinalGetUrl(finalStateVia) {
+  if (!finalStateVia) {
+    if ("${item.request.method}" === "put" && pm.response.code === '201') {
```
Comment (leniatgh) [inline lib/apiScenario/postmanCollectionRunnerClient.ts]: this seems for arm operations only?

File: lib/apiScenario/postmanCollectionRunnerClient.ts [hunk 2]
```diff
@@ -693,19 +704,49 @@ pm.test("Stopped TestProxy recording", function() {
...
+}
+     
 const pollingUrl = pm.response.headers.get("Location") || pm.response.headers.get("Azure-AsyncOperation") || pm.response.headers.get("Operation-Location");
 if (pollingUrl) {
     pm.variables.set("x_polling_url", ${
       this.opts.testProxy
         ? `pollingUrl.replace("${baseUri}","${this.opts.testProxy}")`
         : "pollingUrl"
     });
+    pm.variables.set("x_final_get_url",getLroFinalGetUrl(${finalStateVia}))
```
Comment (leniatgh) [inline lib/apiScenario/postmanCollectionRunnerClient.ts]: prettier format

Comment (leniatgh) [reply lib/apiScenario/postmanCollectionRunnerClient.ts]: host in the url need to be replaced with `testProxy` url.

File: lib/apiScenario/postmanCollectionRunnerClient.ts [hunk 3]
```diff
@@ -909,20 +951,27 @@ try {
...
       },
       baseUri
     );
-    item.request.url = url;
+
+    /**
+     * set for original uri
+     */
+    if (finalStateVia === "original-uri" || (prevMethod === "put" && !finalStateVia)) {
+      item.request.url = url;
```
Comment (leniatgh) [inline lib/apiScenario/postmanCollectionRunnerClient.ts]: can you correlate the initial request and final response, and feed into schema validator? The final response body must match swagger definition per ARM RPC

**Review by leniatgh — COMMENTED** [review]
Looks great. Left some comments, please check.

File: lib/apiScenario/postmanCollectionRunnerClient.ts [hunk 4]
```diff
@@ -933,10 +989,33 @@ try {
...
     }
-    const postScripts = this.generatePostScripts(scriptTypes, undefined, armTemplate);
-    if (postScripts.length > 0) {
-      PostmanHelper.addEvent(item.events, "test", postScripts);
+    let responseAssertion: any;
+    if (finalStateVia === "azure-async-operation") {
+      scriptTypes.push("ResponseDataAssertion");
+      responseAssertion = {
+        "200": [{ test: "/body/properties", expression: "to.be.not.undefined" }],
+      };
```
Comment (leniatgh) [inline lib/apiScenario/postmanCollectionRunnerClient.ts]: Could you define an errorCode for each test? and also update here: https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/api-scenario/references/ErrorCodeReference.md

Comment (jianyexi) [reply lib/apiScenario/postmanCollectionRunnerClient.ts]: Seems this errorCode is for schema validator or example validator, did not find a place to define error code for Api test assertion

File: lib/apiScenario/postmanCollectionRunnerClient.ts (hunk 3 above)
Comment (leniatgh) [reply lib/apiScenario/postmanCollectionRunnerClient.ts]: Corresponding ARM RPC rule: https://github.com/Azure/oav/issues/932

File: lib/apiScenario/newmanReportValidator.ts [hunk 5]
```diff
@@ -234,6 +256,11 @@ export class NewmanReportValidator {
           ? await this.liveValidator.validateLiveRequestResponse(payload)
           : undefined;
 
+        liveValidationForLroFinalGetResult =
+          this.opts.skipValidation && lroFinalPayLoad
```
Comment (leniatgh) [inline lib/apiScenario/newmanReportValidator.ts]:
```suggestion
          !this.opts.skipValidation && lroFinalPayLoad
```

Comment (leniatgh) [reply lib/apiScenario/newmanReportValidator.ts]: this validation only applies to management plane

File: lib/apiScenario/postmanCollectionRunnerClient.ts [hunk 6]
```diff
@@ -693,19 +704,51 @@ pm.test("Stopped TestProxy recording", function() {
...
     if (this.opts.skipLroPoll) return;
-
+    const url = item.request.url;
+    const urlStr = `${baseUri}${url.getPathWithQuery()}`;
     postScripts.push(
       `
-const pollingUrl = pm.response.headers.get("Location") || pm.response.headers.get("Azure-AsyncOperation") || pm.response.headers.get("Operation-Location");
+ const isArmResourceCreate = "${
+   isManagementPlane ?? false
+ }" === "true" && "${item.request.method.toLowerCase()}" === "put" && (pm.response.code === 201 || pm.response.code === 200)
```
Comment (leniatgh) [inline lib/apiScenario/postmanCollectionRunnerClient.ts]: can we instead skip generating certain code for data-plane? for code readability

File: lib/apiScenario/postmanCollectionRunnerClient.ts [hunk 7]
```diff
@@ -783,6 +826,12 @@ try {
       });
     }
 
+    PostmanHelper.addEvent(pollerItem.events, "prerequest", [
+      `
+    console.log(pm.variables.get("x_polling_url"))
```
Comment (leniatgh) [inline lib/apiScenario/postmanCollectionRunnerClient.ts]: we can add `verbose` option condition for this

File: lib/apiScenario/postmanCollectionRunnerClient.ts (hunk 4 above)
Comment (jianyexi) [reply lib/apiScenario/postmanCollectionRunnerClient.ts]: Send https://github.com/Azure/azure-rest-api-specs/pull/21959 to update the errorCode docs

File: lib/apiScenario/postmanCollectionRunnerClient.ts (hunk 7 above)
Comment (jianyexi) [reply lib/apiScenario/postmanCollectionRunnerClient.ts]: this is unused, will remove it

File: lib/apiScenario/newmanReportValidator.ts [hunk 8]
```diff
@@ -173,14 +174,34 @@ export class NewmanReportValidator {
...
+      let lroFinalPayLoad;
+
+      // associate  final get response to the
+      if (it.annotation.type === "LRO" && [200, 201, 202].includes(it.response.statusCode)) {
+        const lroFinalGetExecution = this.getLROFinalResponse(
+          newmanReport.executions,
+          it.annotation.step
+        );
+        if (lroFinalGetExecution?.response.statusCode === 200) {
+          lroFinalPayLoad = this.convertToLROLiveValidationPayload(it, lroFinalGetExecution);
```
Comment (leniatgh) [inline lib/apiScenario/newmanReportValidator.ts:186]: can this final result be added to the example generation?

File: lib/apiScenario/postmanCollectionRunnerClient.ts [hunk 9]
```diff
@@ -716,27 +764,68 @@ pm.test("Stopped TestProxy recording", function() {
...
+    // by default, the final url header is Location for ARM, Operation-Location for dataplane.
+    return ${
+      isArmResourceCreate
+        ? `'${urlStr}'`
+        : `pm.response.headers.get("Location") || pm.response.headers.get("Operation-Location")`
+    }
+  }
+  switch (finalStateVia) {
+    case "location": {
+      return pm.response.headers.get("Location");
```
Comment (leniatgh) [inline lib/apiScenario/postmanCollectionRunnerClient.ts:803]: do we assert the existence of the expected header?

File: lib/apiScenario/postmanCollectionRunnerClient.ts [hunk 10]
```diff
@@ -716,27 +764,68 @@ pm.test("Stopped TestProxy recording", function() {
...
+        isArmResourceCreate ? `'${urlStr}'` : "''"
+      }
 if (pollingUrl) {
-    pm.variables.set("x_polling_url", ${
-      this.opts.testProxy
-        ? `pollingUrl.replace("${baseUri}","${this.opts.testProxy}")`
-        : "pollingUrl"
-    });
+    pm.variables.set("x_polling_url",getProxyUrl(pollingUrl));
+    pm.variables.set("x_final_get_url",getProxyUrl(getLroFinalGetUrl("${finalStateVia ?? ""}")))
```
Comment (leniatgh) [inline lib/apiScenario/postmanCollectionRunnerClient.ts]: please format the code snippet

File: lib/apiScenario/newmanReportValidator.ts [hunk 11]
```diff
@@ -236,6 +258,11 @@ export class NewmanReportValidator {
           ? await this.liveValidator.validateLiveRequestResponse(payload)
           : undefined;
 
+        liveValidationForLroFinalGetResult =
```
Comment (jordan13game) [inline lib/apiScenario/newmanReportValidator.ts:274]: This result needs render in markdown and html report

**Comment by leniatgh** [comment]
Please validate test-proxy works well https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/api-scenario/how-to/MakeTestProxyRecording.md#make-recording

File: lib/apiScenario/postmanCollectionRunnerClient.ts (hunk 9 above)
Comment (jianyexi) [reply lib/apiScenario/postmanCollectionRunnerClient.ts:803]: Yes, in the pre check of the final get

**Review by leniatgh — APPROVED** [review]
Looks good, please update Changelog and bump version

**Pull request merged**